src/neo-vm/ReferenceCounter.cs

```diff
@@ -80,19 +80,18 @@ internal int CheckZeroReferred()
                     HashSet<CompoundType> toBeDestroyedInLoop = new(ReferenceEqualityComparer.Instance);
                     Queue<CompoundType> toBeChecked = new();
                     toBeChecked.Enqueue(compound);
-                    while (toBeChecked.Count > 0)
+                    while (toBeChecked.TryDequeue(out var c))
                     {
-                        CompoundType c = toBeChecked.Dequeue();
                         counter.TryGetValue(c, out Entry? entry);
                         if (entry?.StackReferences > 0)
                         {
                             toBeDestroyedInLoop.Clear();
                             break;
                         }
-                        toBeDestroyedInLoop.Add(c);
+                        if (!toBeDestroyedInLoop.Add(c)) continue;
                         if (entry?.ObjectReferences is null) continue;
                         foreach (var pair in entry.ObjectReferences)
-                            if (pair.Value > 0 && !toBeDestroyed.Contains(pair.Key) && !toBeDestroyedInLoop.Contains(pair.Key))
```

Comment:
@shargon Why modify this line?

Reply:
Because it will be checked again in line 91, in my tests was faster

Reply:
Line 91 checks `c`, but this line checks `pair.Key`. They are different.

Reply:
But `c` comes from `toBeChecked` that it will be the same, we can test it with this `Contains`, maybe it's solved too.

Reply:
It seems solved to, could you check it @vang1ong7ang ?

Reply:
@shargon , @erikzhang  imo with line 91 applied, `!toBeDestroyedInLoop.Contains(pair.Key)` is not necessary but it is not recommended to remove it.

Reply:
However, imo, @shargon 's proposal <https://github.com/neo-project/neo/issues/2528#issuecomment-877139951> is a good way to be considered.

Reply:
As far as I am concerned, the variables added to`toBeChecked` should be added to `toBeDestroyedInLoop` immediately. 
When we want to enqueue anyone, we should check if it has already been enqueued before.

Actually, we're writing a BFS here. We'd better follow the [standard way](https://en.wikipedia.org/wiki/Breadth-first_search) to implement it. 

```
 1  procedure BFS(G, root) is
 2      let Q be a queue
 3      label root as explored     // put to toBeDestroyedInLoop immediately
 4      Q.enqueue(root)
 5      while Q is not empty do
 6          v := Q.dequeue()
 7          if v is the goal then  // our own operations here
 8              return v
 9          for all edges from v to w in G.adjacentEdges(v) do
10              if w is not labeled as explored then  // check it first
11                  label w as explored // mark it immediately to toBeDestroyedInLoop 
12                  Q.enqueue(w)  
```

Reply:
https://github.com/neo-project/neo-vm/pull/416/commits/ef740f61eb275112ee19383b751e400dbc6f20e9 could solve https://github.com/neo-project/neo/issues/2532

Reply:
just did a test, ef740f6 changes the VM behavior i think, it runs unnormally